pkg/router/metrics/haproxy/haproxy.go

```diff
@@ -0,0 +1,648 @@
+package haproxy
+
+import (
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	_ "net/http/pprof"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+)
+
+const (
+	namespace = "haproxy" // For Prometheus metrics.
+
+	// HAProxy 1.4
+	// # pxname,svname,qcur,qmax,scur,smax,slim,stot,bin,bout,dreq,dresp,ereq,econ,eresp,wretr,wredis,status,weight,act,bck,chkfail,chkdown,lastchg,downtime,qlimit,pid,iid,sid,throttle,lbtot,tracked,type,rate,rate_lim,rate_max,check_status,check_code,check_duration,hrsp_1xx,hrsp_2xx,hrsp_3xx,hrsp_4xx,hrsp_5xx,hrsp_other,hanafail,req_rate,req_rate_max,req_tot,cli_abrt,srv_abrt,
+	// HAProxy 1.5
+	// pxname,svname,qcur,qmax,scur,smax,slim,stot,bin,bout,dreq,dresp,ereq,econ,eresp,wretr,wredis,status,weight,act,bck,chkfail,chkdown,lastchg,downtime,qlimit,pid,iid,sid,throttle,lbtot,tracked,type,rate,rate_lim,rate_max,check_status,check_code,check_duration,hrsp_1xx,hrsp_2xx,hrsp_3xx,hrsp_4xx,hrsp_5xx,hrsp_other,hanafail,req_rate,req_rate_max,req_tot,cli_abrt,srv_abrt,comp_in,comp_out,comp_byp,comp_rsp,lastsess,
+	expectedCsvFieldCount = 52
+	statusField           = 17
+
+	frontendType = "0"
+	backendType  = "1"
+	serverType   = "2"
+	listenerType = "3"
+)
+
+var (
+	frontendLabelNames = []string{"frontend"}
+	backendLabelNames  = []string{"backend", "namespace", "route"}
+	serverLabelNames   = []string{"server", "namespace", "route", "pod", "service"}
+)
+
+func newFrontendMetric(metricName string, docString string, constLabels prometheus.Labels) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace:   namespace,
+			Name:        "frontend_" + metricName,
+			Help:        docString,
+			ConstLabels: constLabels,
+		},
+		frontendLabelNames,
+	)
+}
+
+func newBackendMetric(metricName string, docString string, constLabels prometheus.Labels) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace:   namespace,
+			Name:        "backend_" + metricName,
+			Help:        docString,
+			ConstLabels: constLabels,
+		},
+		backendLabelNames,
+	)
+}
+
+func newServerMetric(metricName string, docString string, constLabels prometheus.Labels) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace:   namespace,
+			Name:        "server_" + metricName,
+			Help:        docString,
+			ConstLabels: constLabels,
+		},
+		serverLabelNames,
+	)
+}
+
+type metrics map[int]*prometheus.GaugeVec
+
+func (m metrics) Names() []int {
+	keys := make([]int, 0, len(m))
+	for k := range m {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
+	return keys
+}
+
+// defaultSelectedMetrics is the list of metrics included by default
+var defaultSelectedMetrics = []int{2, 4, 5, 7, 8, 9, 13, 14, 17, 21, 24, 33, 35, 40, 43}
```

Comment:
but if the columns re-order, or new columns are added, we'll break if we don't go by names.